File: android/app/src/main/java/com/ethosprotocol/services/NotificationDeliveryLog.kt
package com.ethosprotocol.services

import android.content.Context
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Local, on-device record of notification lifecycle events (scheduled /
 * delivered / suppressed), so a support ticket like "I never got my TTL
 * warning" is answerable from a debug screen instead of requiring backend
 * log correlation (#235).
 *
 * Doubles as the dedup registry for #232: a `vault_expired`/`vault_released`
 * event applied via the WebSocket is recorded here, and `PushService`'s
 * `onMessageReceived` checks it before showing a banner for the same event.
 *
 * **No sensitive vault data is ever recorded** — only a vault ID, an event
 * type string (e.g. "vault_expired", "ttl_warning"), a delivery channel, and
 * a timestamp. Never balance, beneficiary, or any other vault field.
 */
@Serializable
data class NotificationDeliveryEvent(
    val kind: NotificationDeliveryLog.Kind,
    val source: NotificationDeliveryLog.Source,
    val eventType: String,
    val vaultId: String,
    val timestampMillis: Long
)

@Singleton
class NotificationDeliveryLog @Inject constructor(@ApplicationContext context: Context) {

    enum class Kind { SCHEDULED, DELIVERED, SUPPRESSED }
    enum class Source { LOCAL, PUSH, WEBSOCKET }

    companion object {
        private const val PREFS_NAME = "notification_delivery_log"
        private const val KEY_EVENTS = "events"
        // Bounded so this never grows unbounded across a long-lived install —
        // only recent history is useful for support triage.
        private const val MAX_ENTRIES = 200
    }

    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    @Synchronized
    fun record(kind: Kind, source: Source, eventType: String, vaultId: String, atMillis: Long = System.currentTimeMillis()) {
        val events = loadLocked().toMutableList()
        events.add(NotificationDeliveryEvent(kind, source, eventType, vaultId, atMillis))
        val trimmed = if (events.size > MAX_ENTRIES) events.takeLast(MAX_ENTRIES) else events
        saveLocked(trimmed)
    }

    /** All logged events, most recent first, for the debug screen. */
    @Synchronized
    fun recentEvents(): List<NotificationDeliveryEvent> = loadLocked().reversed()

    @Synchronized
    fun clear() {
        prefs.edit().remove(KEY_EVENTS).apply()
    }

    /**
     * #232: was [eventType] for [vaultId] already delivered via the WebSocket
     * within [windowMillis] of [nowMillis]? Used to suppress a duplicate push
     * banner for an event the UI already reflects.
     */
    @Synchronized
    fun wasRecentlyDeliveredViaWebSocket(
        vaultId: String,
        eventType: String,
        windowMillis: Long = 30_000,
        nowMillis: Long = System.currentTimeMillis()
    ): Boolean = loadLocked().any { event ->
        event.source == Source.WEBSOCKET &&
            event.kind == Kind.DELIVERED &&
            event.vaultId == vaultId &&
            event.eventType == eventType &&
            (nowMillis - event.timestampMillis) in 0..windowMillis
    }

    private val listSerializer = ListSerializer(NotificationDeliveryEvent.serializer())

    private fun loadLocked(): List<NotificationDeliveryEvent> {
        val raw = prefs.getString(KEY_EVENTS, null) ?: return emptyList()
        return runCatching { Json.decodeFromString(listSerializer, raw) }.getOrDefault(emptyList())
    }

    private fun saveLocked(events: List<NotificationDeliveryEvent>) {
        prefs.edit().putString(KEY_EVENTS, Json.encodeToString(listSerializer, events)).apply()
    }
}

File: ios/EthosProtocol/Sources/Services/NotificationDeliveryLog.swift
import Foundation

/// Local, on-device record of notification lifecycle events (scheduled /
/// delivered / suppressed), so a support ticket like "I never got my TTL
/// warning" is answerable from a debug screen instead of requiring backend
/// log correlation (#235).
///
/// Doubles as the dedup registry for #232: a `vault_expired`/`vault_released`
/// event applied via the WebSocket is recorded here, and the push-delivery
/// path (`NotificationService`'s `willPresent`) checks it before showing a
/// banner for the same event.
///
/// **No sensitive vault data is ever recorded** — only a vault ID, an event
/// type string (e.g. `"vault_expired"`, `"ttl_warning"`), a delivery
/// channel, and a timestamp. Never balance, beneficiary, or any other vault
/// field.
struct NotificationDeliveryEvent: Codable, Equatable, Identifiable {
    enum Kind: String, Codable { case scheduled, delivered, suppressed }
    enum Source: String, Codable { case local, push, websocket }

    var id = UUID()
    let kind: Kind
    let source: Source
    /// e.g. "vault_expired", "vault_released", "check_in_reminder", "ttl_warning".
    /// Never raw vault content.
    let eventType: String
    /// Vault ID only.
    let vaultID: String
    let timestamp: Date
}

final class NotificationDeliveryLog {
    static let shared = NotificationDeliveryLog()

    /// Bounded so this never grows unbounded across a long-lived install —
    /// only recent history is useful for support triage.
    private let maxEntries = 200
    private let userDefaultsKey = "com.ethosprotocol.notification_delivery_log"
    private let defaults: UserDefaults
    private let queue = DispatchQueue(label: "com.ethosprotocol.notification-delivery-log")

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    func record(kind: NotificationDeliveryEvent.Kind, source: NotificationDeliveryEvent.Source,
                eventType: String, vaultID: String, at timestamp: Date = Date()) {
        let event = NotificationDeliveryEvent(kind: kind, source: source, eventType: eventType,
                                               vaultID: vaultID, timestamp: timestamp)
        queue.sync {
            var all = loadLocked()
            all.append(event)
            if all.count > maxEntries {
                all.removeFirst(all.count - maxEntries)
            }
            saveLocked(all)
        }
    }

    /// All logged events, most recent first, for the debug screen.
    func recentEvents() -> [NotificationDeliveryEvent] {
        queue.sync { loadLocked().reversed() }
    }

    func clear() {
        queue.sync { defaults.removeObject(forKey: userDefaultsKey) }
    }

    /// #232: was `eventType` for `vaultID` already delivered via the WebSocket
    /// within `window` seconds of `now`? Used to suppress a duplicate push
    /// banner for an event the UI already reflects.
    func wasRecentlyDeliveredViaWebSocket(vaultID: String, eventType: String,
                                          within window: TimeInterval = 30, now: Date = Date()) -> Bool {
        queue.sync {
            loadLocked().contains { event in
                event.source == .websocket
                    && event.kind == .delivered
                    && event.vaultID == vaultID
                    && event.eventType == eventType
                    && now.timeIntervalSince(event.timestamp) <= window
                    && now.timeIntervalSince(event.timestamp) >= 0
            }
        }
    }

    private func loadLocked() -> [NotificationDeliveryEvent] {
        guard let data = defaults.data(forKey: userDefaultsKey),
              let decoded = try? JSONDecoder().decode([NotificationDeliveryEvent].self, from: data) else {
            return []
        }
        return decoded
    }

    private func saveLocked(_ events: [NotificationDeliveryEvent]) {
        guard let data = try? JSONEncoder().encode(events) else { return }
        defaults.set(data, forKey: userDefaultsKey)
    }
}
